--- update.cpp	2004-10-17 12:59:48.000000000 +0200
+++ update.cpp.yadi	2004-08-26 19:55:02.000000000 +0200
@@ -75,10 +75,10 @@
 #ifdef SQUASHFS
 #define LIST_OF_UPDATES_LOCAL_FILENAME "squashfs.list"
 #define UPDATE_LOCAL_FILENAME          "update.squashfs"
-#define RELEASE_CYCLE                  "1.8"
+#define RELEASE_CYCLE                  "2.0"
 #define FILEBROWSER_UPDATE_FILTER      "squashfs"
-#define MTD_OF_WHOLE_IMAGE             5
-#define MTD_DEVICE_OF_UPDATE_PART      "/dev/mtd/3"
+#define MTD_OF_WHOLE_IMAGE             4
+#define MTD_DEVICE_OF_UPDATE_PART      "/dev/mtd/1"
 #else
 #define LIST_OF_UPDATES_LOCAL_FILENAME "cramfs.list"
 #define UPDATE_LOCAL_FILENAME          "update.cramfs"
