--- decoder.cpp.orig	2004-07-12 09:22:06.000000000 +0200
+++ decoder.cpp	2004-09-23 20:43:02.000000000 +0200
@@ -148,28 +148,82 @@
 }
 
 void Decoder::Pause( int flags )
-// flags & 1 = disableAudio
-// flags & 2 = onlyFreeze
-{
-	eDebug("Decoder::Pause()");
-	if (fd.video != -1)
-	{
-		if ( ::ioctl(fd.video, VIDEO_FREEZE) < 0 )
-			eDebug("VIDEO_FREEZE failed (%m)");
-		if ( flags & 2 )
-			return;
-		if ( fd.audio != -1 )
-		{
-			if ( ::ioctl(fd.audio, AUDIO_SET_AV_SYNC, 0) < 0 )
-				eDebug("AUDIO_SET_AV_SYNC failed (%m)");
-			if ( flags & 1 )
-			{
-				if ( ::ioctl(fd.audio, AUDIO_SET_MUTE, 1 )<0)
-					eDebug("AUDIO_SET_MUTE failed (%m)");
-			}
-		}
-	}
-}
+// changes by trans $Id: decoder.cpp.diff,v 1.1 2004/10/08 23:59:19 essu Exp $
+ // flags & 1 = disableAudio 
+ // flags & 2 = onlyFreeze 
+ { 
+    eDebug("Decoder::Pause()"); 
+    if (fd.video != -1) 
+    { 
+       if ( ::ioctl(fd.video, VIDEO_FREEZE) < 0 ) 
+          eDebug("VIDEO_FREEZE failed (%m)"); 
+       if ( flags & 2 ) 
+          return; 
+       if ( fd.audio != -1 ) 
+       { 
+ //trans play ts 
+ #if HAVE_DVB_API_VERSION >= 3 
+             if (::ioctl(fd.audio, AUDIO_STOP)<0) 
+                eDebug("AUDIO_STOP failed(%m)"); 
+             else 
+                eDebug("audio_pause (success)"); 
+ #else 
+ //trans play ts 
+          if ( ::ioctl(fd.audio, AUDIO_SET_AV_SYNC, 0) < 0 ) 
+             eDebug("AUDIO_SET_AV_SYNC failed (%m)"); 
+          if ( flags & 1 ) 
+          { 
+             if ( ::ioctl(fd.audio, AUDIO_SET_MUTE, 1 )<0) 
+                eDebug("AUDIO_SET_MUTE failed (%m)"); 
+          } 
+ //trans play ts 
+ #endif 
+ //trans play ts 
+       } 
+    } 
+ } 
+ 
+ void Decoder::Resume(bool enableAudio) 
+ { 
+    eDebug("Decoder::Resume()"); 
+    if (fd.video != -1) 
+    { 
+ //      clearScreen(); 
+ //trans play ts 
+ #if HAVE_DVB_API_VERSION >= 3 
+       if (::ioctl(fd.video, VIDEO_STOP)<0) 
+          eDebug("VIDEO_STOP failed(%m)"); 
+       if (::ioctl(fd.video, VIDEO_PLAY)<0) 
+          eDebug("VIDEO_PLAY failed(%m)"); 
+ #else 
+ //trans play ts 
+       if (::ioctl(fd.video, VIDEO_CONTINUE)<0) 
+          eDebug("VIDEO_CONTINUE failed(%m)"); 
+       if ( ::ioctl(fd.audio, AUDIO_SET_AV_SYNC, 1 ) < 0 ) 
+          eDebug("AUDIO_SET_AV_SYNC failed (%m)"); 
+ //trans play ts 
+ #endif 
+ //trans play ts 
+       if ( enableAudio )  // Video Clip Mode 
+       { 
+ //trans play ts 
+ #if HAVE_DVB_API_VERSION >= 3 
+          if (::ioctl(fd.audio, AUDIO_PLAY)<0) 
+             eDebug("AUDIO_PLAY failed (%m)"); 
+          else 
+             eDebug("audio continue (success)"); 
+ #else 
+ //trans play ts 
+          if (::ioctl(fd.audio, AUDIO_SET_MUTE, 0 )<0) 
+             eDebug("AUDIO_SET_MUTE failed (%m)"); 
+          else 
+             eDebug("audio_pause (success)"); 
+ //trans play ts 
+ #endif 
+ //trans play ts 
+       } 
+    } 
+ } 
 
 void Decoder::Resume(bool enableAudio)
 {
