--- rules-archive.org	2004-08-11 00:37:35.668260464 +0200
+++ rules-archive	2004-08-11 00:37:17.180071096 +0200
@@ -10,7 +10,7 @@
 # core
 
 u-boot-1.1.1.tar.bz2;http://heanet.dl.sourceforge.net/sourceforge/u-boot
-linux-2.4.25.tar.bz2;ftp://ftp.de.kernel.org/pub/linux/kernel/v2.4;ftp://ftp.kernel.org/pub/linux/kernel/v2.4
+linux-2.4.26.tar.bz2;ftp://ftp.de.kernel.org/pub/linux/kernel/v2.4;ftp://ftp.kernel.org/pub/linux/kernel/v2.4
 cifs-1.20c-2.4.tar.gz;http://us1.samba.org/samba/ftp/cifs-cvs
 linux-2.4.25-cifs24.103.diff.bz2;http://tuxbox.berlios.de/pub/tuxbox/cdk/src
 linux-2.4.25-dbox2-mtd.diff.bz2;http://tuxbox.berlios.de/pub/tuxbox/cdk/src
