--- dbox2.h	2004-01-13 19:50:10.000000000 +0000
+++ dbox2.h.new	2003-10-29 15:30:27.000000000 +0000
@@ -27,6 +27,9 @@
  * board/config.h - configuration options, board specific
  */
 
+#define	__CONFIG_FLASH	1
+#define	__CONFIG_JFFS2_FLASH	1
+
 #ifndef __CONFIG_H
 #define __CONFIG_H
 
@@ -50,27 +53,14 @@
 #undef	CONFIG_8xx_CONS_SMC2
 #undef	CONFIG_8xx_CONS_NONE
 
-#if 0
-#define CONFIG_BOOTDELAY	-1	/* autoboot disabled		*/
-#else
-#define CONFIG_BOOTDELAY	1	/* autoboot after 5 seconds	*/
-#endif
+#define CONFIG_BOOTDELAY	0	/* autoboot 		*/
 
 #undef	CONFIG_BOOTARGS
 
 #define	CONFIG_BOOTCOMMAND							\
-	"dhcp; tftp \"$(bootfile)\"; "						\
-	"protect off 10040000 107fffff; "					\
-        "setenv bootargs root=/dev/nfs rw nfsroot=$(serverip):$(rootpath) "	\
-	"ip=$(ipaddr):$(serverip):$(gatewayip):$(netmask):$(hostname)::off "	\
-	"console=$(console); "							\
+	"fsload; setenv bootargs root=/dev/mtdblock2 rw console=$(console); "	\
 	"bootm"
-/*
-#define	CONFIG_BOOTCOMMAND							\
-	"protect off 10040000 107fffff; "					\
-	"fsload; setenv bootargs root=/dev/mtdblock2 console=$(console); "	\
-	"bootm"
-*/
+
 
 #define	CONFIG_EXTRA_ENV_SETTINGS 						\
 	"console=ttyS0\0"
@@ -84,30 +74,27 @@
 
 #define	CONFIG_COMMANDS		( CONFIG_CMD_DFL | CFG_CMD_FS | CFG_CMD_DHCP )
 
-#define	CONFIG_FS		( CFG_FS_CRAMFS | CFG_FS_JFFS2 )
+#define	CONFIG_FS		( CFG_FS_JFFS2 )
 
-#define	CFG_FS_PART0_TYPE	CFG_FS_CRAMFS
+#define	CFG_FS_PART0_TYPE	CFG_FS_JFFS2
 #define	CFG_FS_PART0_OFFSET	0x10040000
-#define	CFG_FS_PART0_SIZE	0x6e0000
-#define	CFG_FS_PART1_TYPE	CFG_FS_JFFS2
-#define	CFG_FS_PART1_OFFSET	0x10720000
-#define	CFG_FS_PART1_SIZE	0xe0000
+#define	CFG_FS_PART0_SIZE	0x7c0000
 
-#define	CONFIG_DBOX2_FS_ENV_READ		"1:env"
+#define	CONFIG_DBOX2_ENV_READ_FS		"0:var/tuxbox/boot/ppcboot.conf"
 
 #define	CONFIG_TUXBOX_NETWORK			1
 
 #ifdef	CONFIG_LCD_BOARD
 #define	CONFIG_DBOX2_LCD_INFO			1
 #define	CONFIG_DBOX2_LCD_LOGO			1
-#define	CONFIG_DBOX2_LCD_LOGO_FS		"1:logo-lcd"
+#define	CONFIG_DBOX2_LCD_LOGO_FS		"0:boot/logo-lcd"
 #define	CONFIG_DBOX2_LCD_LOGO_TFTP		"logo-lcd"
 #define	CONFIG_DBOX2_LCD_LOGO_RESERVE		2
 #undef	CONFIG_DBOX2_LCD_FONT8x16
 #endif	/* CONFIG_LCD_BOARD */
 #ifdef	CONFIG_DBOX2_FB
 #define	CONFIG_DBOX2_FB_LOGO			1
-#define	CONFIG_DBOX2_FB_LOGO_FS			"1:logo-fb"
+#define	CONFIG_DBOX2_FB_LOGO_FS			"0:boot/logo-fb"
 #define	CONFIG_DBOX2_FB_LOGO_TFTP		"logo-fb"
 #endif	/* CONFIG_DBOX2_FB */
 
