--- dbox2-flash.c.ori	2004-08-29 18:58:21.000000000 +0200
+++ dbox2-flash.c	2004-08-29 19:02:46.000000000 +0200
@@ -32,14 +32,8 @@
 	.mask_flags	= 0
 	},
 	{
-	.name		= "root (cramfs)",	
-	.size		= 7040 * 1024, 
-	.offset		= MTDPART_OFS_APPEND, 
-	.mask_flags	= 0
-	},
-	{
-	.name		= "var (jffs2)",
-	.size		= 896 * 1024, 
+	.name		= "root (jffs2)",	
+	.size		= 7936 * 1024, 
 	.offset		= MTDPART_OFS_APPEND, 
 	.mask_flags	= 0
 	},
