--- rules-make.old	2004-05-04 20:06:06.000000000 +0200
+++ rules-make	2004-05-04 20:08:26.000000000 +0200
@@ -17,7 +17,7 @@
 binutils;2.14.90.0.5;binutils-2.14.90.0.5;binutils-2.14.90.0.5.tar.bz2;extract:binutils-2.14.90.0.5.tar.bz2
 bootstrap_gcc;3.3.3;gcc-3.3.3;gcc-core-3.3.3.tar.bz2:gcc-core.diff;extract:gcc-core-3.3.3.tar.bz2;patch:gcc-core.diff
 gcc;3.3.3;gcc-3.3.3;gcc-core-3.3.3.tar.bz2:gcc-g++-3.3.3.tar.bz2;extract:gcc-core-3.3.3.tar.bz2;extract:gcc-g++-3.3.3.tar.bz2
-glibc;2.3.2;glibc-2.3.2;glibc-2.3.2.tar.bz2:glibc-linuxthreads-2.3.2.tar.bz2:glibc-2.3.2.diff;extract:glibc-2.3.2.tar.bz2;dirextract:glibc-linuxthreads-2.3.2.tar.bz2;patch:glibc-2.3.2.diff
+glibc;2.2.5;glibc-2.2.5;glibc-2.2.5.tar.bz2:glibc-linuxthreads-2.2.5.tar.bz2:glibc.diff;extract:glibc-2.2.5.tar.bz2;dirextract:glibc-linuxthreads-2.2.5.tar.bz2;patch:glibc.diff
 uclibc;0.9.11;uClibc-0.9.11;uClibc-0.9.11.tar.bz2:uclibc-0.9.11.diff;extract:uClibc-0.9.11.tar.bz2;patch:uclibc-0.9.11.diff
 watchdog;5.2.4;watchdog-5.2.4.orig;watchdog_5.2.4.orig.tar.gz:watchdog.diff;extract:watchdog_5.2.4.orig.tar.gz;patch:watchdog.diff
 
