--- u-boot.2x_flash.dbox2.h.ori	2004-05-03 13:40:10.000000000 +0200
+++ u-boot.2x_flash.dbox2.h	2004-05-03 13:49:54.000000000 +0200
@@ -50,15 +50,12 @@
 #undef	CONFIG_8xx_CONS_SMC2
 #undef	CONFIG_8xx_CONS_NONE
 
-#if 0
-#define CONFIG_BOOTDELAY	-1	/* autoboot disabled		*/
-#else
-#define CONFIG_BOOTDELAY	1	/* autoboot after 5 seconds	*/
-#endif
+#define CONFIG_BOOTDELAY	0	/* autoboot after 5 seconds	*/
+
 
 #undef	CONFIG_BOOTARGS
 #define	CONFIG_BOOTCOMMAND							\
-	"setenv bootargs console=$(console) root=/dev/mtdblock2; protect off 10020000 107fffff; fsload; bootm"
+	"setenv bootargs console=$(console) root=/dev/mtdblock2 rw; protect off 10020000 107fffff; fsload; bootm"
 
 #define	CONFIG_EXTRA_ENV_SETTINGS 						\
 	"console=ttyS0\0"
@@ -72,29 +69,26 @@
 
 #define	CONFIG_COMMANDS		( CONFIG_CMD_DFL | CFG_CMD_FS | CFG_CMD_DHCP )
 
-#define	CONFIG_FS		( CFG_FS_CRAMFS | CFG_FS_JFFS2 )
+#define	CONFIG_FS		( CFG_FS_JFFS2 )
 
-#define	CFG_FS_PART0_TYPE	CFG_FS_CRAMFS
+#define	CFG_FS_PART0_TYPE	CFG_FS_JFFS2
 #define	CFG_FS_PART0_OFFSET	0x10040000
-#define	CFG_FS_PART0_SIZE	0x6e0000
-#define	CFG_FS_PART1_TYPE	CFG_FS_JFFS2
-#define	CFG_FS_PART1_OFFSET	0x10720000
-#define	CFG_FS_PART1_SIZE	0xe0000
+#define	CFG_FS_PART0_SIZE	0x7c0000
 
-#define	CONFIG_DBOX2_ENV_READ_FS		"1:tuxbox/boot/boot.conf"
+#define	CONFIG_DBOX2_ENV_READ_FS		"0:var/tuxbox/boot/boot.conf"
 
 #define	CONFIG_TUXBOX_NETWORK			1
 
 #ifdef	CONFIG_LCD_BOARD
 #define	CONFIG_DBOX2_LCD_INFO			1
 #define	CONFIG_DBOX2_LCD_LOGO			1
-#define	CONFIG_DBOX2_LCD_LOGO_FS		"1:tuxbox/boot/logo-lcd"
+#define	CONFIG_DBOX2_LCD_LOGO_FS		"0:var/tuxbox/boot/logo-lcd"
 #define	CONFIG_DBOX2_LCD_LOGO_RESERVE		2
 #undef	CONFIG_DBOX2_LCD_FONT8x16
 #endif	/* CONFIG_LCD_BOARD */
 #ifdef	CONFIG_DBOX2_FB
 #define	CONFIG_DBOX2_FB_LOGO			1
-#define	CONFIG_DBOX2_FB_LOGO_FS			"1:tuxbox/boot/logo-fb"
+#define	CONFIG_DBOX2_FB_LOGO_FS			"0:var/tuxbox/boot/logo-fb"
 #endif	/* CONFIG_DBOX2_FB */
 
 #define	CONFIG_AUTOBOOT_SELECT			1
